### PR TITLE
Honor WooCommerce Settings When Formatting Currency/Price

### LINKED ISSUE #1090: Honor WooCommerce Settings When Formatting Currency/Price
**Is your feature request related to a problem? Please describe.**

The current `formatCurrency()` function uses the store locale, but not the specific currency formatting options like "currency position". While it's unlikely that these will differ, it is possible.

> I think there could be some confusion about site language controlling the position/formatting of the symbol and separators rather than the WooCommerce > Settings > Currency Options. Could we roll our own format function that is a js version of `wc_price`/`get_woocommerce_price_format`?
> -- @justinshreve https://github.com/woocommerce/wc-admin/pull/122#pullrequestreview-130841773

**Describe the solution you'd like**

Use the WooCommerce Currency settings when formatting price values in JavaScript.

**Should this be prioritized? Why?**

Yeah, @timmyc saw this as the next step:

> I say we get this merged in so we at least have some proper currency formatting in place, and make an issue to possibly try using WC currency settings down the road.
> -- @timmyc https://github.com/woocommerce/wc-admin/pull/122#pullrequestreview-131807939

### LINKED ISSUE #1322: Honor user locale/WC settings when formatting chart Y-axis numbers
Discussed in: https://github.com/woocommerce/wc-admin/pull/1312#discussion_r248012285

Currently, chart Y-axis numbers are always displayed in the [default format](https://github.com/woocommerce/wc-admin/blob/master/packages/components/src/chart/index.js#L29-L34): `1,300,300.54`, but they should use the user locale or, even better, the WooCommerce settings defined in `WooCommerce > Settings > Currency Options` when it's a currency.

Related issue: #1090.

### PR DESCRIPTION
**Pull request by jeffstieler**
Fixes #1090.
Fixes #1322.

This PR seeks to format numbers and prices using the store's format settings (WooCommerce > Settings > Currency Options). It makes use of a JS port of PHP's `number_format()` since the existing JS options are lacking the ability to specify separator characters outside of providing a locale.

In addition to reworking `numberFormat()` and `formatCurrency()`, this PR:
* Adds currency symbol to individual Orders retrieved through the REST API
* HTML entity decodes currency symbol and price format (to avoid using `dangerouslySetInnerHTML`)

### Screenshots

screen shot 2019-01-17 at 10 15 53 am

### Detailed test instructions:

- Set currency options that don't match the locale of the store (e.g. position="right with space", decimal sep=",", thousand_sep=".", precision="3" for USD)
- Verify that all numbers and prices use this format in reports, including the Y-axis of charts

### CODE REVIEW

**Review by timmyc — COMMENTED** [review]
This is great to see! We already had some feedback around currency formatting on one of the release posts, so this will save quite a bit of bug reports over the coming months - thanks Jeff!

Had a few questions in the code, but overall looking fantastic. Could you also add a note to CHANGELOG.md in the `components` and `currency` packages folder? Since we just released them yesterday, you'll probably have to start new minor version sections like `1.4.1 ( unreleased ) in https://github.com/woocommerce/wc-admin/blob/master/packages/components/CHANGELOG.md

**Review by Aljullu — COMMENTED** [review]
Closing two issues in one PR. :heart_eyes: 

This is testing well for me, couldn't find any number that is not correctly formatted. I just added a couple of comments inline with the code.

File: client/lib/number/index.js [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
+const number_format = require( 'locutus/php/strings/number_format' );
+
 /**
  * Formats a number using site's current locale
  *
- * @format
- * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
- * @param {Number|String}  number    number to format
- * @returns {?String}                  A formatted string.
+ * @see https://locutus.io/php/strings/number_format/
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply client/lib/number/index.js]: I'll swap to the HTTP version.

File: client/lib/number/index.js [hunk 2]
```diff
@@ -1,20 +1,46 @@
...
+			0,
+			decimalSeparator,
+			thousandSeparator
+		);
+
+		if ( decimals ) {
+			return formattedNumber + decimalSeparator + decimals;
+		}
+
+		return formattedNumber;
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply client/lib/number/index.js]: Nice! Updated in d2cb3f77.

File: client/lib/number/index.js [hunk 3]
```diff
@@ -1,20 +1,46 @@
...
+
 	if ( isNaN( number ) ) {
 		return '';
 	}
-	return new Intl.NumberFormat( locale ).format( number );
+
+	const decimalSeparator = get( wcSettings, [ 'currency', 'decimal_separator' ], '.' );
+	const thousandSeparator = get( wcSettings, [ 'currency', 'thousand_separator' ], ',' );
+
+	if ( null === precision ) {
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply client/lib/number/index.js]: Update in d2cb3f77.

File: client/lib/number/test/index.js [hunk 4]
```diff
@@ -16,4 +16,16 @@ describe( 'numberFormat', () => {
 	it( 'should accept a string', () => {
 		expect( numberFormat( '10000' ) ).toBe( '10,000' );
 	} );
+
+	it( 'maintains all decimals if no precision specified', () => {
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply client/lib/number/test/index.js]: Added in 48249ea0.

File: packages/currency/src/index.js [hunk 5]
```diff
@@ -3,28 +3,35 @@
...
-export function formatCurrency( number, currency ) {
-	const locale = wcSettings.siteLocale || 'en-US'; // Default so we don't break.
-	// default to wcSettings if currency is not passed in
-	if ( ! currency ) {
-		currency = get( wcSettings, 'currency.code', 'USD' );
-	}
-	if ( 'number' !== typeof number ) {
-		number = parseFloat( number );
+export function formatCurrency( number, currencySymbol ) {
+	// default to wcSettings if currency symbol is not passed in
```
[the remark this answers is not in the document]
Comment (jeffstieler) [reply packages/currency/src/index.js]: Done in e32dedad.

**Review by jeffstieler — COMMENTED** [review]
> Could you also add a note to CHANGELOG.md in the components and currency packages folder?

Done in 8f0d27c6. This ready for another review.

**Review by timmyc — APPROVED** [review]
Thanks for the updates, great refactor suggestion @Aljullu - this is still testing great for me.

**Pull request merged**